### DISCUSSION TITLE
Build Assistant Teams

### DISCUSSION THREAD

**Discussion by ashpreetbedi** [agno-agi/agno#66] [Ideas]
Assistants should be able to delegate tasks to other Assistants. The Assistant `Team` should have:
- Leader (the main Assistant that does the delegation)
- Assistants (Assistants the do the task)
- Reviewer (Assistant that validates the work before returning to the user)

ETA: mid-march

**Comment by kinthaiofficial** [comment]
Building assistant teams in phidata is a great use case — a few patterns that make it work reliably in production:

**Role specialization over generalization**
Resist the urge to give every assistant access to every tool. A "research assistant" should have search + read tools. A "writing assistant" should have draft + edit tools. A "code assistant" should have execute + debug tools. Specialization makes each assistant more predictable and easier to evaluate.

**Coordinator vs worker pattern**
Have a lightweight coordinator assistant whose only job is to route tasks to the right specialist. The coordinator doesn't do the work — it decides who does. This avoids the "one assistant does everything and loses track" problem for complex multi-step tasks.

**Shared memory, private scratch space**
All assistants in a team should read from a shared knowledge base (your RAG store, session context) but write to their own scratch space first. The coordinator then decides what gets committed to shared context. This prevents one assistant's intermediate state from polluting the others.

**Explicit handoff messages**
When assistant A finishes and hands off to assistant B, the handoff message should include: what was done, what was decided, what constraints were applied. Don't rely on shared context alone — explicit handoffs make debugging much easier.

**Async task queues over sync chains**
If task A must complete before task B starts, you're building a fragile sync chain. Better: task A posts its result to a shared queue; task B polls for it. This decouples failure modes and makes individual retries clean.

What's the team topology you're building? Flat (all assistants coordinate directly) or hierarchical (coordinator → specialists)?

**Comment by kinthaiofficial** [comment]
Building Assistant Teams in Phidata is one of its most powerful patterns — here's a working example for a research + writing team:

```python
from phi.agent import Agent
from phi.model.openai import OpenAIChat
from phi.tools.duckduckgo import DuckDuckGo
from phi.tools.newspaper4k import Newspaper4k

# Specialized agents
web_researcher = Agent(
    name="Web Researcher",
    model=OpenAIChat(id="gpt-4o"),
    tools=[DuckDuckGo(), Newspaper4k()],
    instructions=[
        "Search for the most recent and relevant information",
        "Read full articles when URLs are found",
        "Summarize key findings with citations",
    ],
    show_tool_calls=True,
)

writer = Agent(
    name="Content Writer",
    model=OpenAIChat(id="gpt-4o"),
    instructions=[
        "Write clear, well-structured content based on research",
        "Include specific facts and cite sources inline",
        "Use headers and bullet points for readability",
    ],
)

editor = Agent(
    name="Editor",
    model=OpenAIChat(id="gpt-4o"),
    instructions=[
        "Review content for accuracy and clarity",
        "Ensure consistent tone and style",
        "Add a brief TL;DR summary at the top",
    ],
)

# Create the team
team = Agent(
    team=[web_researcher, writer, editor],
    model=OpenAIChat(id="gpt-4o"),
    instructions=[
        "1. Have the Web Researcher gather information",
        "2. Have the Writer create initial content",
        "3. Have the Editor review and finalize",
    ],
    show_tool_calls=True,
    markdown=True,
)

# Run the team
team.print_response(
    "Write a comprehensive analysis of AI agent economic models",
    stream=True,
)
```

**Key considerations for agent teams**:
- **Budget allocation**: Each sub-agent call costs tokens. Set `max_tokens` limits per agent.
- **Context passing**: The team agent synthesizes outputs from sub-agents — clarity in instructions matters.
- **Parallelism**: Phidata runs sub-agents sequentially by default; for independent tasks you can use async.

For patterns on agent economic coordination in teams: https://blog.kinthai.ai/agent-wallet-economic-models-autonomous-agents
